# Midterm (MACS 30150), February 11, 2019 [20 points total]
### Solutions
### [Zhihan Yu]
You will have an hour and fifty minutes to take this exam.
1. Put your name in the space provided above (without the brackets).
2. Download this notebook and accompanying data `MidtermScores.csv`.
3. The class will disconnect their wi-fi and begin taking the exam.
4. The exam is open computer, closed internet, open note.
5. Once you are done, you will signal to a TA or instructor to come over and monitor while you reconnect to the internet and upload your edited notebook to Canvas.

Remember that you must clearly indicate your answers to all questions below. We will give partial credit for partially correct answers. The midterm is worth 20 points (2 problem sets).

# 1. Symbolic and Numerical Derivatives [10 points total]
This exercise will use the following function. In a household savings problem, the marginal utility of consumption is an important function. If a household has a constant relative risk aversion utility function, the marginal utility functional form can be the following:
\begin{equation}
  MU(c_t) \equiv \bigl(c_t\bigr)^{-\sigma}
\end{equation}
where $c_t>0$ is current period consumption and $\sigma\geq 1$ is the coefficient of constant relative risk aversion.

Assume the budget constraint every period is $c_t + b_{t+1} = (1+r_t)b_t + w_t$, where $b_t$ is the initial wealth in period $t$, $b_{t+1}$ is the savings chosen in period $t$ that comes back to the household in period $t+1$ with interest, and $w_t$ is the wage paid on the unit of inelastically supplied labor. Then we can rewrite the marginal utility of consumption above with the budget constraint substituted in.
\begin{equation}
  MU = \Bigl([1 + r_t]b_t + w_t - b_{t+1}\Bigr)^{-\sigma}
\end{equation}


## 1a) Plot the MU(c) function for different consumption values [2 points]
Assume that the coefficient of relative risk aversion is $\sigma = 1.8$. Use Python to plot the marginal utility of consumption for 100 equally spaced consumption values between 0.7 and 5.0. Make sure your plot has an $x$-axis labeled "consumption" and a $y$-axis labeled "marginal utility".

In [1]:
# Import dependencies
import numpy as np
from matplotlib import pyplot as plt


# Define sigma
sigma = 1.8

# 100 equally spaced consumptiuon values between 0.7 and 5.0
xvals = np.linspace(0.7, 5.0, 100)

# Define the utility function (lambdified)
def marginalUtilty(c_t):
    return c_t**(-sigma)

# Plot my graph
plt.plot(xvals, marginalUtilty(xvals))
plt.title(r"Graph of marginal utility with $\sigma = 1.8$")
plt.ylabel("marginal uitlity")
plt.xlabel("consumption")
plt.show()

<Figure size 640x480 with 1 Axes>

## 1b) Symbolic numerical derivative [2 points]
Assume the version of the marginal utility above that has the budget constraint substituted in.
\begin{equation}
  MU = \Bigl([1 + r_t]b_t + w_t - b_{t+1}\Bigr)^{-\sigma}
\end{equation}
Use Python's symbolic math package `sympy` to compute the analytical derivative of the marginal utility function with respect to beginning-of-period wealth $b_t$. This will involve you declaring each variable and parameter as a symbol, then specifying the $MU$ funtion to be differentiated. The solution should be a function of all the variables and parameters $(b_t, b_{t+1}, r_t, w_t, \sigma)$. Leave the output in the form that `sympy` produces.

In [2]:
# Import dependency
import sympy as sym

# Define my function as a string... replace b_t and b_{t+1} to b_now and b_next
 ## I renamed variable sigma to sig because there is alreadya global var name sigma.. 
mu_sub_str = '((1+r)*b_now + w - b_next)**(-sig)'

# Since I can treat all my other variables (r, w, b_next) as just variables, I'm only going to deal with declaring b_t 
# as a symbol
derivative = sym.diff(sym.sympify(mu_sub_str), sym.symbols('b_now'))

print(derivative)
print(type(derivative)) # It's a sympy core object

-sig*(r + 1)*(-b_next + b_now*(r + 1) + w)**(-sig)/(-b_next + b_now*(r + 1) + w)
<class 'sympy.core.mul.Mul'>


## 1c) Write a function that produces the analytical derivative [3 points]
Write a function that takes five inputs which are the values for the variables and parameters of the marginal utility $(b_t, b_{t+1}, r_t, w_t, \sigma)$ and returns the analytical derivative of the marginal utility (scalar) with respect to beginning-of-period wealth $b_t$ associated with those values. Show your function definition in the cell below. Name your function `dMU_c`. Print out the value of the function for the following three combinations of variables and parameters.

$$ \frac{\partial MU_1}{\partial b_t}:\quad b_t=1.0, \:\: b_{t+1}=0.8, \:\: r_t=0.04, \:\: w_t=1.2, \:\: \sigma=2.2 $$

$$ \frac{\partial MU_2}{\partial b_t}:\quad b_t=5.0, \:\: b_{t+1}=2.1, \:\: r_t=0.01, \:\: w_t=1.8, \:\: \sigma=2.0 $$

$$ \frac{\partial MU_3}{\partial b_t}:\quad b_t=3.3, \:\: b_{t+1}=3.4, \:\: r_t=0.03, \:\: w_t=1.0, \:\: \sigma=1.8 $$

In [3]:
# Let's first lambdify the derivative.
# Two approaches can be taken: 1) use sympy lambdify (this is very useful for generalized solutions)
#                              2) write out a function that takes in values for the five variables and return the calcualted value
# Since this test is a one-time thing. I'll go with method 2. It's just copy-pasting. 

def dMU_c(b_now, b_next, r, w, sig):
    # copy paste from above output...output[2]
    return (-sig*(r + 1)*(-b_next + b_now*(r + 1) + w)**(-sig)/(-b_next + b_now*(r + 1) + w)) 

print("With the parameters bt=1.0...: ", dMU_c(1.0, 0.8, 0.04, 1.2, 2.2))
print("With the parameters bt=5.0...: ", dMU_c(5.0, 2.1, 0.01, 1.8, 2.0))
print("With the parameters bt=3.3...: ", dMU_c(3.3, 3.4, 0.03, 1.0, 1.8))

With the parameters bt=1.0...:  -0.7123542921746638
With the parameters bt=5.0...:  -0.018848228604752878
With the parameters bt=3.3...:  -1.8592010790841615


## 1d) Compute the numerical derivative approximation [3 points]
Use the centered second-order finite difference approximation of the derivative to calculate the numerical derivative of the marginal utility function.

$$ f'(x_0)\approx \frac{\bigl(f(x_0 + h)\bigr) - \bigl(f(x_0 - h)\bigr)}{2h} $$

Recall that our marginal utility function is a function of four variables and the parameter $\sigma$.
\begin{equation}
  MU = \Bigl([1 + r_t]b_t + w_t - b_{t+1}\Bigr)^{-\sigma}
\end{equation}
So we want to approximate the derivative of the marginal utility with respect to beginning-of-period wealth $\partial MU(c)/\partial b_t$ by using the centered second-order finite difference approximation above. A helpful hint on how to do this is the marginal utility function is like $f(x)$ and the derivative is the centered second-order approximation equation.

Define a function named `dMU_c_approx` that takes as inputs the five variables and parameters $(b_t, b_{t+1}, r_t, w_t, \sigma)$ as well as a finite difference amount $h$. Have this function return the centered second-order approximation of the derivative of the marginal utility function with respect to beginning-of-period wealth.

Using your `dMU_c` function from part (c), print the analytical derivative of the marginal utility function with respect to beginning-of-period wealth with the following parameterization.

$$ \frac{\partial MU}{\partial b_t}:\quad b_t=1.0, \:\: b_{t+1}=0.8, \:\: r_t=0.04, \:\: w_t=1.2, \:\: \sigma=2.2 $$

Use your `dMU_c_approx` function to print out the numerical approximation of the derivative for the specification above with the following three finite difference amounts: $h_1 = 0.1$, $h_2= 0.001$ and  $h_3=0.00001$.

In [4]:
# now it's easier to use the sympy lamdify.
MU_lamb = sym.lambdify([sym.symbols('b_now'), sym.symbols('b_next'), sym.symbols('r'), sym.symbols('w'), \
                        sym.symbols('sig')], sym.sympify(mu_sub_str))

# Centered Second Order Difference Quotient
def dMU_c_approx(values, step):
    """
    This function gives me the centered second approximation of the derivative.
    -----------------
    ARGS: 
        values: array-like of length 5
            this contains the initialization points. The tuple is of the following order
            (b_now, b_next, r, w, sig)
        step: float
            this is the 'h' defined in equation above
    
    RETURN:
        this returns the approximation to the derivative
    """
    
    # Unpack the values
    b_now, b_next, r, w, sig = values
    
    
    # There are two steps in the approximation: an upwarad and downward step
    b_now_stepUp = b_now + step
    b_now_stepDown = b_now - step
    
    # Write out the numerator
    numerator = MU_lamb(b_now_stepUp, b_next, r, w, sig) - MU_lamb(b_now_stepDown, b_next, r, w, sig)
    denom = 2*step
    
    return(numerator/denom)


print("Sympoblic Differentiation solution: ", dMU_c(1.0, 0.8, 0.04, 1.2, 2.2))
print("-------------------------")
print("Centered Second Approximation solution (h=0.1): ", dMU_c_approx((1.0, 0.8, 0.04, 1.2, 2.2), 0.1))
print("Centered Second Approximation solution (h=0.001): ", dMU_c_approx((1.0, 0.8, 0.04, 1.2, 2.2), 0.001))
print("Centered Second Approximation solution (h=0.00001): ", dMU_c_approx((1.0, 0.8, 0.04, 1.2, 2.2), 0.00001))

Sympoblic Differentiation solution:  -0.7123542921746638
-------------------------
Centered Second Approximation solution (h=0.1):  -0.720747904181559
Centered Second Approximation solution (h=0.001):  -0.7123551244865578
Centered Second Approximation solution (h=0.00001):  -0.7123542922538028


# 2. Maximum likelihood estimation [10 points total]
This exercise will make use of the `MidtermScores.csv` dataset, which contains 97 observations of students' midterm scores $scores_i$ as well as the number of hours they studied for the midterm $hours_i$ and whether the midterm was within a week of Chinese New Year $ch\_ny_i$.

* `scores`: score of $i$th student on midterm
* `hours`: number of hours that $i$th student studied for the midterm
* `ch_ny`: =1 if the midterm was within a week of Chinese New Year, =0 otherwise

You can load these data into Python as a numpy array using the `np.loadtxt()` command.

In [5]:
import numpy as np

data = np.loadtxt('MidtermScores.csv', delimiter=',', skiprows=1)
print('Shape of the data=', data.shape)
scores = data[:, 0]
hours = data[:, 1]
ch_ny = data[:, 2]

Shape of the data= (97, 3)


Assume that our model of student midterm scores is the following linear model.

$$ scores_i = \beta_0 + \beta_1 hours_i + \beta_2 ch\_ny_i + \varepsilon_i \quad\text{where}\quad \varepsilon_i\sim N(0,\sigma) $$

where the error $\varepsilon_i$ are normally distributed with mean 0 and standard deviation $\sigma$.

## 2a) Log likelihood function [3 points]
Define a function named `log_lik` that takes as inputs three data vectors $scores_i$, $hours_i$, $ch\_ny_i$, three coefficient parameters $\beta_0$, $\beta_1$, $\beta_2$, and the standard deviation $\sigma$ of the normally distributed errors and returns the log likelihood (a scalar). Print the log likelihood value for the first 10 data observations given all the coefficients are equal to one $\beta_0, \beta_1, \beta_2=1.0$ and the standard deviation of the errors is equal to $\sigma=0.5$. For the answer to this question, I want to see a function and one scalar log-likelihood value.

In [6]:
'''
From the model provided above, we know that the errors are distributed normally with mean 0 and std sigma
Thus, rearranging, we can see

    scores - b0 - b1*hours - b2*ch_ny = error ~ N(0, sig)
'''

# Observe the max score
# print(scores.max())

# Import dependencies
from scipy.stats import norm


def log_lik (scores, hours, ch_ny, b0, b1, b2, sig):
    
    
    # Create my epsilon errors
    epsilons = scores - b0 - b1*hours - b2*ch_ny
    normal_densities = norm.pdf(epsilons, 0, sig)
    ln_densities = np.log(normal_densities)
    log_likeli = np.sum(ln_densities)
    
    return(log_likeli)

print(log_lik(scores[:10], hours[:10], ch_ny[:10], 1., 1., 1., 0.5))

-186.8990404522523


## 2b) Estimate the parameters by MLE [7 points]
Estimate the four parameters $\beta_0, \beta_1, \beta_2, \sigma$ to maximize the log likelihood from part (a). Remember that a minimizer will have to minimize the negative of the log likelihood. Use all 97 observations for this estimation. Use an initial guess of $\beta_0=10.0$, $\beta_1=1.0$, $\beta_2=1.0$, and $\sigma=$ the standard deviation of the $scores_i$ variable. Report your estimated coefficients $\beta_0, \beta_1, \beta_2$ and standard deviation $\sigma$. What is the predicted effect on a midterm score of a student who takes the test around Chineses New Year $ch\_ny_i=1$ versus the score of a student who does not take the test around Chinese New year $ch\_ny_i=0$, other things being equal?

In [7]:
# Import dependencies

from scipy import optimize as opt

# So I've already gotten the log-likelihood. Normally, we wish to maximize this guy, but it's apparently easier to 
# minimize. Thus, we need to appeal to the dual of the optmization problem


def objective_(params, *args):
    b0, b1, b2, sig = params
    score, hour, cn = args
    return(-log_lik(score, hour, cn, b0, b1, b2, sig))


# # Testing the function that Dr.Evans gave us
# test_sample = np.array(scipy.stats.norm.rvs(loc=0, scale=2, size=500000))
# print(test_sample.var())
# # Close enough. But array.std gives me the stdev instead of variance. so i'll go with std


arguments = (scores, hours, ch_ny)
params_init = (10., 1., 1., scores.std())

results_uncstr = opt.minimize(objective_, params_init, arguments)

print(results_uncstr.message)


list_param = ['beta0', 'beta1', 'beta2', 'sigma']
for ind in range(len(results_uncstr.x)):
    print("The MLE estimator for %s is %0.3f." %(list_param[ind], results_uncstr.x[ind]))
    
    
# Update my the regression using the proper coefficients
b0, b1, b2, sig = results_uncstr.x
print("-------------")
print("The regression: ")
print("score = %0.3f + %0.3f * hours + %0.3f * ch_ny + error"%(b0, b1, b2))

Optimization terminated successfully.
The MLE estimator for beta0 is 4.750.
The MLE estimator for beta1 is 0.716.
The MLE estimator for beta2 is -1.891.
The MLE estimator for sigma is 0.653.
-------------
The regression: 
score = 4.750 + 0.716 * hours + -1.891 * ch_ny + error


From the above equation, we see that if it is Chinese New Year (ch_ny=1), we'd see a drop in 1.89 points for the student's scores. In input[6], I observed the maximum score for the dataset, it returned 19.32...Thus, it's logical to assume that the maximum score for the midterm was 20. Thus, a 1.89 drop in the scores is a decrease of 9.45%. That's quite significant. 